File: 0061-rotate-list/0061-rotate-list.java
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return null;
        }

        ListNode last = head;
        int length = 1;

        while (last.next != null) {
            last = last.next;
            length++;
        }

        k = k % length;
        last.next = head;

        for (int i = 0; i < length - k - 1; i++) {
            head = head.next;
        }

        ListNode result = head.next;
        head.next = null;

        return result;
    }
}

File: 0061-rotate-list/0061-rotate-list.py
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or k == 0:
            return head

        node = head
        tail = None
        n = 0

        while node:
            n += 1
            if not node.next:
                tail = node
            node = node.next

        k %= n
        if k == 0:
            return head

        tail.next = head
        node = head

        for i in range(n - k - 1):
            node = node.next

        head = node.next
        node.next = None

        return head